### ISSUE TITLE
New Mastodon API endpoints

### ISSUE THREAD

**Issue by annando** [friendica/friendica#7967]
Here is a list of API endpoints that should be supported next. They do appear in my logfile and they are generated by external statistics tools or clients like Fedilab.

* [x] /api/v1/custom_emojis (#8191)
* [x] /api/v1/notifications?since_id=123456&exclude_types[]=follow&exclude_types[]=favourite&exclude_types[]=reblog&exclude_types[]=poll&limit=30 (#10223)
* [x] /api/v1/timelines/home?limit=20 (#10220)
* [x] /api/v1/timelines/public?local=true&limit=40 (#10220)
* [x] /api/v1/directory?order=active&local=true&limit=20&offset=0 (#9115)
* [x] /api/v1/directory?order=new&local=true&limit=20&offset=0 (#9115)

**Comment by nupplaphil** [comment]
@MrPetovan I'd say I take this one:
> /api/v1/notifications?since_id=123456&exclude_types[]=follow&exclude_types[]=favourite&exclude_types[]=reblog&exclude_types[]=poll&limit=30
during my rework - After #8186 , there will be already an API-Object for it :-)

**Comment by MrPetovan** [comment]
Let me discourage you. The notification endpoint requires:
- the Notification entity, which requires:
    - the Account entity, that exists,
    - the Status entity, which requires:
        - the Attachment entity,
        - the Application entity,
        - the Mention entity,
        - the Tag entity, which requires:
           - the History entity, although it's optional.
        - the Emoji entity that now exists, hence why I started with the custom_emojis endpoint,
        - the Poll entity, but we can't use it yet,
        - the Card entity.

I would suggest tackling any of the timelines endpoints first, because you'll get to create the `Status` entity, and then the notification endpoint will be a breeze. The other way around, you'll run into more trouble than you need.

**Comment by nupplaphil** [comment]
At least it was a try ;-)))
And tbh I'd love doing such API-stuff rather than Frontend things :D ...

**Comment by MrPetovan** [comment]
There are plenty enough of Mastodon API endpoints for us to share. These are only the ones appearing in @annando log file, but if we want 100% coverage, there's a longer way to go.

**Comment by nupplaphil** [comment]
OT: I think the API stuff is REALLY good testable, so before creating a lot of new APIs, we (I? ;-)) should do some changes to introduce Module/API testability :-)

**Comment by MrPetovan** [comment]
The problem that I still see is that all the `rawContent` methods are exiting and if we want them to return something it has to be format-independent between HTML, JSON and XML (RSS or Atom).

**Comment by nupplaphil** [comment]
Json/XML will be possible if we use the System exit methods as dynamic classes. Then we can override their behavior, add a "output" string where we store the output instead of calling header/print/exit :).
HTML will be possible if make the Renderer dynamic and do the same as with the exits ;)

**Comment by MrPetovan** [comment]
Yep, but in the meantime, not sure how to make API endpoints testable. With a separate method that generates the content that `rawContent` calls?

**Comment by stom79** [comment]
I would like to introduce push notifications with the app. Is it possible with the current api to use since_id for pagination? I have already bring the support on my end.

I am using these endpoints:
- `/statuses/mentions_timeline.json`
- `/statuses/retweets_of_me.json`
- `/statuses/followers.json`

**Comment by MrPetovan** [comment]
Unlikely, but it something we can probably add. I’ll create a separate ticket for it.

**Comment by annando** [comment]
@MrPetovan are you still working on this? In the process of removing old stuff ("gcontact" table for example) I want to get rid of the whole "poco" handling. But for this to make possible, we need to have got the ` /api/v1/directory` endpoint.

**Comment by MrPetovan** [comment]
This is still planned, I can prioritize it.

**Comment by annando** [comment]
Would be great. I think that the directory stuff is really simple: https://docs.joinmastodon.org/methods/instance/directory/

**Comment by annando** [comment]
I guess we have to move this issue again to the next release.
